# 01 — Inicijalni pregled skupa podataka

U ovoj svesci vršimo prvi pregled sirovih podataka. Cilj je da se upoznamo sa
strukturom skupa, proverimo tipove podataka i uočimo eventualne probleme
(nedostajuće vrednosti, duplikate, nelogične vrednosti) **pre** bilo kakve obrade.

**Važno:** u ovoj fazi podatke samo *posmatramo i dijagnostikujemo* — ne menjamo
ih i ne čistimo. Čišćenje i obrada slede u kasnijim sveskama, nakon podele na
trening i test skup.

## 1. Učitavanje podataka i biblioteka

Za rad koristimo biblioteku **pandas**, standardni alat za rad sa tabelarnim
podacima u Python-u. Učitavamo skup podataka iz CSV fajla u strukturu koja se
zove **DataFrame** (tabela sa redovima i kolonama).

In [1]:
# Uvozimo biblioteke potrebne za ovu fazu.
# pandas (pd) - za rad sa tabelama.
# numpy (np) - za numeričke operacije; koristićemo ga kasnije za rad sa NaN vrednostima.
import pandas as pd
import numpy as np

# Podešavanje da pandas prikazuje SVE kolone kada ispisujemo tabelu.
# Bez ovoga, pandas skraćuje prikaz i sakriva srednje kolone sa "...".
pd.set_option("display.max_columns", None)

# Učitavamo CSV fajl u DataFrame df.
# "../" znači da izlazimo iz foldera notebooks/ u glavni folder repoa.
df = pd.read_csv("../WA_Fn-UseC_-Telco-Customer-Churn.csv")

print("Podaci su uspešno učitani.")

Podaci su uspešno učitani.


## 2. Dimenzije skupa podataka

Prvo proveravamo osnovne dimenzije — koliko redova (korisnika) i koliko kolona
(atributa) skup sadrži.

In [2]:
# df.shape vraća par (broj_redova, broj_kolona).
print("Broj redova (korisnika):", df.shape[0])
print("Broj kolona (atributa):", df.shape[1])

Broj redova (korisnika): 7043
Broj kolona (atributa): 21


## 3. Prvi pogled na podatke

Prikazujemo prvih nekoliko redova kako bismo videli konkretne vrednosti i stekli
osećaj za sadržaj svake kolone.

In [3]:
# df.head() prikazuje prvih 5 redova tabele.
# Kao poslednja linija u ćeliji, Jupyter je automatski prikazuje kao tabelu.
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


## 4. Informacije o kolonama i tipovima podataka

Funkcija `info()` daje pregled svih kolona: njihov naziv, broj popunjenih
(ne-null) vrednosti i **tip podataka** (`dtype`).

Ovo je ključno za rano otkrivanje problema. Posebno nas zanima da li se tipovi
poklapaju sa očekivanjima — na primer, `TotalCharges` bi trebalo da bude broj,
pa ako je prepoznata kao tekst (`object`), to je znak da nešto nije u redu.

In [4]:
# df.info() ispisuje pregled: naziv kolone, broj ne-null vrednosti i tip podataka.
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

## 5. Provera nedostajućih vrednosti

Standardnu proveru nedostajućih vrednosti radimo funkcijom `isnull()`, koja za
svaku ćeliju vraća `True` ako je vrednost nedostajuća (`NaN`), a `False` inače.
Sabiranjem tih vrednosti po kolonama dobijamo broj nedostajućih vrednosti u svakoj koloni.

In [5]:
# df.isnull() vraća tabelu True/False (True = vrednost nedostaje).
# .sum() sabira po kolonama: koliko True (nedostajućih) ima u svakoj koloni.
print("Broj nedostajućih vrednosti po kolonama:")
print(df.isnull().sum())

Broj nedostajućih vrednosti po kolonama:
customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64


## 6. Skrivene nedostajuće vrednosti

Prethodna provera pokazala je da nema nedostajućih vrednosti — ali to može biti
obmanjujuće. U realnim podacima nedostajuće vrednosti se ponekad kriju kao
**prazni razmaci** (`" "`) umesto pravih `NaN` oznaka. Funkcija `isnull()` ih
ne prepoznaje jer su tehnički tekst, a ne prazne vrednosti.

Zato dodatno proveravamo da li neka kolona sadrži vrednosti koje su samo prazan razmak.

In [6]:
# Poredimo svaku ćeliju sa praznim razmakom " ".
# (df == " ") vraća tabelu True/False (True = ćelija sadrži tačno jedan razmak).
# .sum() sabira po kolonama koliko takvih "praznih" vrednosti ima.
print("Broj vrednosti koje su prazan razmak, po kolonama:")
print((df == " ").sum())

Broj vrednosti koje su prazan razmak, po kolonama:
customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64


## 7. Analiza problema sa kolonom `TotalCharges`

Sada razumemo zašto je `TotalCharges` prepoznata kao tekst (`object`) umesto broja:
pošto 11 vrednosti sadrži prazan razmak (koji nije broj), pandas je celu kolonu
protumačio kao tekstualnu.

Da bismo mogli da analiziramo ovu kolonu numerički, moramo je konvertovati u
broj. Vrednosti koje se ne mogu konvertovati (prazni razmaci) pretvorićemo u
prave `NaN` oznake, koje potom možemo pravilno da analiziramo.

In [7]:
# pd.to_numeric() pokušava da pretvori kolonu u broj.
# errors="coerce" znači: vrednosti koje se NE mogu pretvoriti u broj
# (prazni razmaci) postaju NaN umesto da izazovu grešku.
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")

# Ponovo proveravamo nedostajuće vrednosti — sada bi TotalCharges trebalo da
# pokaže 11 NaN vrednosti (bivši prazni razmaci).
print("Broj nedostajućih vrednosti nakon konverzije TotalCharges:")
print(df.isnull().sum())

Broj nedostajućih vrednosti nakon konverzije TotalCharges:
customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64


**Ovde zakljucujemo da TotalCharges bi trebalo da bude `numericka vrednost` sto nam govori da je treba da je konvertujemo**. To mozemo uraditi preko ugradjene funkcije u biblioteci `pandas`. Vrednosti koje ne mogu da se konvertuju postaju `NaN` vrednosti.

In [8]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"],errors="coerce")

## 8. Analiza korisnika sa nedostajućim `TotalCharges`

Nakon konverzije `TotalCharges` u numerički tip, sada možemo da vidimo tačno
koliko korisnika ima nedostajuću vrednost. Detaljnijom analizom želimo da
razumemo *zašto* baš ti korisnici imaju problem — da li je u pitanju slučajna
greška ili postoji neki obrazac.

In [9]:
print("Nedostajuće vrednosti nakon konverzije TotalCharges:")
print(df.isnull().sum())
print()


Nedostajuće vrednosti nakon konverzije TotalCharges:
customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64



### Ključna opservacija

Svih 11 korisnika sa nedostajućom vrednošću `TotalCharges` ima **`tenure = 0`**.
To znači da se radi o korisnicima koji su tek započeli korišćenje usluge i za
koje u trenutku formiranja skupa podataka **nije postojala evidentirana ukupna naplata**.

Ovo je ključan uvid za dalje odluke:
- Nije reč o slučajnoj grešci merenja
- Nije reč o namerno preskočenom polju
- Vrednost logički **ne postoji** za te korisnike (nova pretplata, još nije bilo naplate)

**Odluku o rukovanju sa ovim redovima donećemo u kasnijoj svesci**, nakon
podele na trening i test skup — kako bismo izbegli curenje informacija
(*data leakage*).

Ovde dolazimo do zakljucka da je `postoje` nedostajucih vrednosti u nasem skupu podataka i u sledecem koraku da utvrdimo koje su to vrednosti

In [10]:
print("Korisnici sa nedostajućim TotalCharges:")
print(df[df["TotalCharges"].isnull()])
print()

Korisnici sa nedostajućim TotalCharges:
      customerID  gender  SeniorCitizen Partner Dependents  tenure  \
488   4472-LVYGI  Female              0     Yes        Yes       0   
753   3115-CZMZD    Male              0      No        Yes       0   
936   5709-LVOEQ  Female              0     Yes        Yes       0   
1082  4367-NUYAO    Male              0     Yes        Yes       0   
1340  1371-DWPAZ  Female              0     Yes        Yes       0   
3331  7644-OMVMY    Male              0     Yes        Yes       0   
3826  3213-VVOLG    Male              0     Yes        Yes       0   
4380  2520-SGTTA  Female              0     Yes        Yes       0   
5218  2923-ARZLG    Male              0     Yes        Yes       0   
6670  4075-WKNIU  Female              0     Yes        Yes       0   
6754  2775-SEFEE    Male              0      No        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  \
488            No  No phone service        

## 9. Provera duplikata

Proveravamo da li skup sadrži duplirane redove, kao i da li se neki `customerID`
pojavljuje više puta. Duplikati mogu iskriviti kasnije analize i modeliranje.

In [11]:
print("Broj dupliranih redova:", df.duplicated().sum())
print("Broj dupliranih customerID vrednosti:",df["customerID"].duplicated().sum())
print()


Broj dupliranih redova: 0
Broj dupliranih customerID vrednosti: 0



Zakljucujemo da `nema duplikata` u nasem skupu podataka

## 10. Osnovna statistika numeričkih promenljivih

Za numeričke kolone (`SeniorCitizen`, `tenure`, `MonthlyCharges`, `TotalCharges`)
funkcija `describe()` prikazuje ključne statističke pokazatelje: broj vrednosti,
prosek, standardnu devijaciju, minimum, maksimum i kvartile.

Ovo nam pomaže da:
- Uočimo raspone vrednosti (npr. najveću/najmanju platu)
- Otkrijemo eventualne nelogičnosti (negativne vrednosti kod atributa koji
  ne mogu biti negativni)
- Steknemo osnovni osećaj za raspodelu podataka pre detaljne analize u
  narednim sveskama

In [12]:
print("Osnovna statistika numeričkih promenljivih:")
print(df[["SeniorCitizen", "tenure","MonthlyCharges", "TotalCharges"]].describe())
print()

Osnovna statistika numeričkih promenljivih:
       SeniorCitizen       tenure  MonthlyCharges  TotalCharges
count    7043.000000  7043.000000     7043.000000   7032.000000
mean        0.162147    32.371149       64.761692   2283.300441
std         0.368612    24.559481       30.090047   2266.771362
min         0.000000     0.000000       18.250000     18.800000
25%         0.000000     9.000000       35.500000    401.450000
50%         0.000000    29.000000       70.350000   1397.475000
75%         0.000000    55.000000       89.850000   3794.737500
max         1.000000    72.000000      118.750000   8684.800000



### Tumačenje osnovne statistike

Iz prikazanih vrednosti izdvajamo nekoliko važnih zapažanja:

- **`tenure`**: opseg od 0 do 72 meseca (do 6 godina). Vrednost 0 se odnosi na
  onih 11 korisnika sa nedostajućim `TotalCharges` koje smo prethodno uočili.
- **`MonthlyCharges`**: opseg od ~18 do ~119 dolara. Nema negativnih ni
  besmislenih vrednosti.
- **`TotalCharges`**: opseg od ~18 do ~8.685 dolara. Broj vrednosti je 7032
  (a ne 7043) jer 11 redova ima `NaN` — u skladu sa prethodnom analizom.
- **`SeniorCitizen`**: iako je prikazana kao numerička (vrednosti 0 i 1),
  ova kolona zapravo predstavlja **binarnu kategoriju**. `describe()` na njoj
  ima ograničenu smislenost i za nju je bolja provera raspodele kategorija.

## 11. Provera kategorijskih promenljivih

### 11.1 SeniorCitizen — binarna kategorija u brojnom obliku

Kolona `SeniorCitizen` je zapisana kao broj (0 ili 1), ali predstavlja
**binarnu kategoriju**: da li je korisnik stariji građanin ili ne.
Umesto statistike, prikazujemo raspodelu po vrednostima da vidimo koliko
korisnika pripada svakoj grupi.

In [13]:
print("Raspodela SeniorCitizen:")
print(df["SeniorCitizen"].value_counts())
print()

Raspodela SeniorCitizen:
SeniorCitizen
0    5901
1    1142
Name: count, dtype: int64



Većina korisnika (~5.900, ~84%) nisu stariji građani, dok ih je manjina
(~1.140, ~16%) u toj kategoriji. Ova nesrazmera je bitna informacija za
kasnije analize — istražićemo da li starost korisnika utiče na odliv.

Izdvajamo sve kolone koje imaju `tekstualne` vrednosti

Za svaku kategoriju prikazujemo sve *`postojece`* kategorije i *`broj`* korisnika u svakoj kategoriji

### 11.2 Sve kategorijske promenljive

Prolazimo kroz sve kolone tekstualnog tipa (`object`) i za svaku prikazujemo
postojeće kategorije zajedno sa brojem korisnika u svakoj. Ovo nam omogućava da:

- Vidimo tačne vrednosti koje se javljaju u svakoj koloni
- Uočimo eventualne neočekivane ili nekonzistentne kategorije
- Steknemo osećaj o balansu između kategorija (npr. da li dominira jedna vrednost)

Napomena: kolonu `customerID` treba izuzeti jer sadrži 7043 jedinstvenih
identifikatora — njihov prikaz nema analitičku vrednost.

In [14]:
# Izdvajamo sve kolone tekstualnog tipa.
categorical_columns = df.select_dtypes(include="object").columns

# Za svaku kategorijsku kolonu prikazujemo raspodelu.
# Izuzimamo customerID jer predstavlja jedinstveni identifikator korisnika,
# a ne kategoriju za analizu.
for column in categorical_columns:
    if column == "customerID":
        continue
    print(f"\n{column}:")
    print(df[column].value_counts())


gender:
gender
Male      3555
Female    3488
Name: count, dtype: int64

Partner:
Partner
No     3641
Yes    3402
Name: count, dtype: int64

Dependents:
Dependents
No     4933
Yes    2110
Name: count, dtype: int64

PhoneService:
PhoneService
Yes    6361
No      682
Name: count, dtype: int64

MultipleLines:
MultipleLines
No                  3390
Yes                 2971
No phone service     682
Name: count, dtype: int64

InternetService:
InternetService
Fiber optic    3096
DSL            2421
No             1526
Name: count, dtype: int64

OnlineSecurity:
OnlineSecurity
No                     3498
Yes                    2019
No internet service    1526
Name: count, dtype: int64

OnlineBackup:
OnlineBackup
No                     3088
Yes                    2429
No internet service    1526
Name: count, dtype: int64

DeviceProtection:
DeviceProtection
No                     3095
Yes                    2422
No internet service    1526
Name: count, dtype: int64

TechSupport:
TechSupport
No   

Churn
No     5174
Yes    1869
Name: count, dtype: int64


/tmp/ipykernel_764/3554621375.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = df.select_dtypes(include="object").columns


### Tumačenje kategorijskih promenljivih

**Balansirane binarne kolone:**
- `gender`: Male 3555, Female 3488 — praktično podjednako
- `Partner`: No 3641, Yes 3402 — blaga prevaga onih bez partnera

**Nebalansirane binarne kolone:**
- `Dependents`: No 4933, Yes 2110 — većina korisnika nema zavisne osobe
- `PhoneService`: Yes 6361, No 682 — ogromna većina ima telefonsku uslugu
- `PaperlessBilling`: Yes 4171, No 2872

**Kolone sa tri vrednosti (`Yes` / `No` / `"No internet service"` ili slično):**
`MultipleLines`, `OnlineSecurity`, `OnlineBackup`, `DeviceProtection`,
`TechSupport`, `StreamingTV`, `StreamingMovies`.

> **Važno:** vrednosti `"No internet service"` i `"No"` **nisu isto**:
> - `"No"` znači da korisnik ima internet, ali ne koristi tu dodatnu uslugu
> - `"No internet service"` znači da korisnik uopšte nema internet
>
> Ova razlika biće bitna za feature engineering — odluka o tome da li ih
> spojiti ili razdvojiti donećemo u kasnijoj svesci.

**Kolone sa više kategorija:**
- `InternetService`: Fiber optic 3096, DSL 2416, No 1520
- `Contract`: Month-to-month 3875 (dominira), One year 1472, Two year 1685
- `PaymentMethod`: Electronic check 2365, Mailed check 1604, Bank transfer 1542, Credit card 1521

**Ciljna promenljiva `Churn`**:
- No: 5173 (73%)
- Yes: 1869 (27%)

Klase su **nebalansirane** — što će imati direktan uticaj na izbor metrika
i strategije modeliranja. Detaljna analiza sledi u kasnijim sveskama.

---

## 12. Zaključak

Kroz inicijalni pregled skupa `Telco Customer Churn` došli smo do sledećih
saznanja:

**Struktura skupa:**
- 7.043 korisnika × 21 kolona
- Kolone su konceptualno podeljene u 4 grupe: demografske, usluge,
  informacije o nalogu i plaćanju, i ciljna promenljiva

**Uočeni problemi (za rešavanje u narednim sveskama):**
1. **Skrivene nedostajuće vrednosti u `TotalCharges`** — 11 korisnika (0.16%)
   sa `tenure = 0`. Nakon konverzije u numerički tip, ovi redovi imaju `NaN`.
   Odluku o njihovom uklanjanju donosimo posle podele na trening/test skup.
2. **`SeniorCitizen` u brojnom obliku** — iako je binarna kategorija, zapisana
   je kao 0/1. Razmotrićemo konverziju u `No`/`Yes` radi konzistentnosti.
3. **Vrednosti `"No internet service"` i `"No phone service"`** — nisu isto
   što i `"No"`. Odluka o njihovom tretmanu ide u fazu pripreme podataka.
4. **Nebalansirane klase u `Churn`** — ~73% No, ~27% Yes. Značajno utiče na
   izbor metrika i pristupa modeliranju.

**Pozitivni nalazi:**
- Nema duplikata (ni potpunih ni po `customerID`)
- Nema nelogičnih vrednosti u numeričkim kolonama (negativne, van opsega)
- Sve kategorijske kolone imaju konzistentne vrednosti (bez tipfelera)

